src/middleware/RetryMiddleware.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Middleware, MiddlewareContext, MiddlewareResult } from './types'
+
+export interface RetryOptions {
+  maxRetries: number
+  retryOnTimeout: boolean
+  retryBackoff: (min: number, max: number, attempt: number) => number
+}
+
+export class RetryMiddleware implements Middleware {
```

Comment:
I wonder if there's a way to encapsulate [all this retry logic](https://github.com/elastic/elastic-transport-js/blob/43f061da04d7d9183845bda9b8e693caea3f2ce5/src/Transport.ts#L599-L701) in the middleware so we can pull it out of `Transport`. Improves the readability of `Transport.request` and probably makes testing retry logic a lot easier, too.